cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/AttachmentPullProcessor.java

```diff
@@ -0,0 +1,46 @@
+package com.cloudant.sync.replication;
+
+import com.cloudant.mazha.CouchClient;
+import com.cloudant.sync.datastore.Attachment;
+import com.cloudant.sync.datastore.AttachmentException;
+import com.cloudant.sync.datastore.PreparedAttachment;
+import com.cloudant.sync.datastore.UnsavedStreamAttachment;
+
+import java.io.InputStream;
+
+public class AttachmentPullProcessor implements CouchClient
+        .InputStreamProcessor<PreparedAttachment> {
+
+    private final DatastoreWrapper datastoreWrapper;
+    private final String name;
+    private final String contentType;
+    private final Attachment.Encoding encoding;
+    private final long length;
+    private final long encodedLength;
+
+    AttachmentPullProcessor(DatastoreWrapper wrapper, String name, String contentType, String
+            encoding, long length, long encodedLength) {
+        this.datastoreWrapper = wrapper;
+        this.name = name;
+        this.contentType = contentType;
+        this.encoding = Attachment.getEncodingFromString(encoding);
+        this.length = length;
+        this.encodedLength = encodedLength;
+    }
+
+    static PreparedAttachment pulLAttachmentWithRetry(CouchClient client, DatastoreWrapper
+            wrapper, String id, String rev, String name, String contentType, String encoding,
+                                                      long length, long encodedLength) {
+        return client.processAttachmentStream(id, rev, name, true, new AttachmentPullProcessor
+                (wrapper, name, contentType, encoding, length, encodedLength));
+    }
+
+    @Override
+    public PreparedAttachment processStream(InputStream stream) throws AttachmentException {
+        UnsavedStreamAttachment usa = new UnsavedStreamAttachment(stream, name, contentType,
+                encoding);
+        PreparedAttachment attachment = datastoreWrapper.prepareAttachment(usa, length,
```

Comment:
@tomblench I've had a go at improving this in 293446e - I moved the static method into the `CouchDB` interface (and parameterized it to instead accept an `InputStreamProcessor<T>`) with the implementation in the `CouchClientWrapper`.

This has these advantages:
1. Avoiding the need to cast the `CouchDB` in the `PullStrategy` to a `CouchClientWrapper` to get access to the `CouchClient` it can be used as a `CouchDB` as was the case previously.
1. De-coupling the `DatastoreWrapper` from the `CouchClient` in the `AttachmentPullProcessor`. This makes the `PullStrategy` the only place where the coupling between the `CouchDB` and `DatastoreWrapper` need exist (which is where the coupling was before).

I also fixed the typo :)
WDYT?